File: kalmanFilter.py
import numpy as np

def kalmanFilter(z_c, dz, r, z_r, r_c, r_c_old, p, hessian, numSensors):
    s = np.array([z_c, *dz])  # 3x1
    m = 0.001 * np.identity(3)  # 3x1
    R = 0.001 * np.identity(numSensors)  # 4x4
    u = 0.001 * np.identity(4)  # 4x4
    a = np.array([[1, *(r_c - r_c_old)],
                  [0, 1, 0],
                  [0, 0, 1]])  # 3x3
    h = np.array([
        0,
        *(hessian @ (r_c - r_c_old))  # 2x2 . 2x1 = 2x1
    ])  # 3x1
    hc = np.array([hessian[0, 0], hessian[1, 0], hessian[0, 1], hessian[1, 1]])  # 4x1
    c = np.hstack((np.ones((numSensors, 1)), r - r_c))  # 4x3
    d = 0.5 * np.vstack([np.kron(pt - r_c, pt - r_c) for pt in r])  # 4x4


    s_e = a @ s + h  # 3x1
    p_e = a @ p @ a.T + m  # 3x3 @ 3x1 @ 3x3 = 3x3

    # c @ p_e @ c.T = 4x3 @ 3x3 @ 3x4 = 4x4
    # d @ u @ d.T = 4x4 @ 4x4 @ 4x4 = 4x4
    # p_e @ c.T @ inv... = 3x3 @ 3x4 @ 4x4 = 3x4

    K = p_e @ c.T @ np.linalg.inv((c @ p_e @ c.T) + (d @ u @ d.T) + R)  # 3x4

    # c @ s_e = 4x3 @ 3x1 = 4x1
    # d @ hc = 4x4 @ 4x1 = 4x1
    # K @ ... = 3x4 @ 4x1 = 3x1
    s = s_e + K @ (z_r - (c @ s_e) - (d @ hc))  # 3x1

    # d @ u @ d.T = 4x4 @ 4x4 @ 4x4 = 4x4
    # c.T @ .. @ c = 3x4 @ 4x4 @ 4x3 = 3x3
    p = np.linalg.inv(np.linalg.inv(p_e) + c.T @ np.linalg.inv(d @ u @ d.T + R) @ c)  # 3x3

    z_c = s[0]
    dz = s[1:]
    return z_c, dz, p

File: kalmanFilter.m
function [z_c,Dz,P] = kalmanFilter(z_c,Dz,r,z_r,r_c,r_c_old,P,Hessian,SensorN)
%SensorN = SensorN - 1;%why minus 1?

 s = [z_c,Dz].';
M = 0.001 * eye(3); R = 0.001 * eye(SensorN); 
 U = 0.001 * eye(4);%TODO: change size of U 
A = [1 (r_c - r_c_old).';zeros(2,1) eye(2)]; %A_k_s
h = [0;Hessian * (r_c - r_c_old)]; %h(k-1)
 %P = 0.0001 * zeros(3)
 
Hc=[Hessian(:,1).' Hessian(:,2).'].';

for i = 1:SensorN
    C(i,:) = [1 (r(:,i) - r_c).'];
    D(i,:) = 0.5 * kron((r(:,i)-r_c),(r(:,i)-r_c));
end


    s_e = A * s + h;
    P_e = A * P * A.' + M;
    K = P_e * C.' *  inv(C*P_e*C.'+ D*U*D.' + R);
    
    s = s_e + K * (z_r(1:SensorN).' - C*s_e - D*Hc);

    P = inv(inv(P_e) + C.' * inv(D*U*D.' + R) * C); % why P is needed since we already have s ?
    %P = (eye(3) - K*C) * P_e;
z_c = s(1); Dz = s(2:end)';
z_c = double(z_c);
Dz = double(Dz);

